File: subversion/bindings/javahl/src/org/apache/subversion/javahl/types/DiffOptions.java
/*
 * ====================================================================
 *    Licensed to the Apache Software Foundation (ASF) under one
 *    or more contributor license agreements.  See the NOTICE file
 *    distributed with this work for additional information
 *    regarding copyright ownership.  The ASF licenses this file
 *    to you under the Apache License, Version 2.0 (the
 *    "License"); you may not use this file except in compliance
 *    with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing,
 *    software distributed under the License is distributed on an
 *    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *    KIND, either express or implied.  See the License for the
 *    specific language governing permissions and limitations
 *    under the License.
 * ====================================================================
 */

package org.apache.subversion.javahl.types;

/**
 * Options that control the output of {@link ISVNClient#diff}.
 * @since 1.8
 */
public class DiffOptions
{
    public enum Flag
    {
        /** Ignore all white space */
        IgnoreWhitespace  (0x01),

        /** Ignore changes in amount of white space */
        IgnoreSpaceChange (0x02),

        /** Ignore changes in EOL style */
        IgnoreEOLStyle    (0x04),

        /** Show C function name */
        ShowFunction      (0x08),

        /** Use git's extended diff format */
        GitFormat        (0x10),

        /** Show 'svn:mergeinfo' in human-readable form */
        FormatMergeinfo  (0x20);

        final int value;

        private Flag (int value)
        {
            this.value = value;
        }
    };

    /**
     * @param flagset any combination of the Flag enumeration values.
     */
    public DiffOptions(Flag... flagset)
    {
        int f = 0;
        for (Flag flag : flagset)
        {
            f |= flag.value;
        }
        this.flags = f;
    }

    /** @return whether IgnoreWhitespace is enabled */
    public boolean getIgnoreWhitespace()
    {
        return (0 != (flags & Flag.IgnoreWhitespace.value));
    }

    /** @return whether IgnoreSpaceChange is enabled */
    public boolean getIgnoreSpaceChange()
    {
        return (0 != (flags & Flag.IgnoreSpaceChange.value));
    }

    /** @return whether IgnoreEOLStyle is enabled */
    public boolean getIgnoreEOLStyle()
    {
        return (0 != (flags & Flag.IgnoreEOLStyle.value));
    }

    /** @return whether ShowFunction is enabled */
    public boolean getShowFunction()
    {
        return (0 != (flags & Flag.ShowFunction.value));
    }

    /** @return whether GitFormat is enabled */
    public boolean getGitFormat()
    {
        return (0 != (flags & Flag.GitFormat.value));
    }

    /** @return whether FormatMergeinfo is enabled */
    public boolean getFormatMergeinfo()
    {
        return (0 != (flags & Flag.FormatMergeinfo.value));
    }

    private final int flags;
}

File: subversion/bindings/javahl/native/DiffOptions.cpp
/**
 * @copyright
 * ====================================================================
 *    Licensed to the Apache Software Foundation (ASF) under one
 *    or more contributor license agreements.  See the NOTICE file
 *    distributed with this work for additional information
 *    regarding copyright ownership.  The ASF licenses this file
 *    to you under the Apache License, Version 2.0 (the
 *    "License"); you may not use this file except in compliance
 *    with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing,
 *    software distributed under the License is distributed on an
 *    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *    KIND, either express or implied.  See the License for the
 *    specific language governing permissions and limitations
 *    under the License.
 * ====================================================================
 * @endcopyright
 *
 * @file DiffOptions.cpp
 * @brief Implementation of the DiffOptions class
 */

#include "DiffOptions.h"

namespace {
static jint get_diff_options_flags(jobject joptions)
{
  if (!joptions)
    return 0;

  JNIEnv *const env = JNIUtil::getEnv();

  static volatile jfieldID fid = 0;
  if (!fid)
    {
      fid = env->GetFieldID(env->GetObjectClass(joptions), "flags", "I");
      if (JNIUtil::isJavaExceptionThrown())
        return 0;
    }

  const jint flags = env->GetIntField(joptions, fid);
  if (JNIUtil::isJavaExceptionThrown())
    return 0;
  return flags;
}
} // anonymous namespace

DiffOptions::DiffOptions(jobject joptions)
  : flags(get_diff_options_flags(joptions))
{}


apr_array_header_t *DiffOptions::optionsArray(SVN::Pool &resultPool) const
{
  // JavaHL ignores the default diff options from the client config
  // files, so we always have to allocate an array, even if it will
  // remain empty.
  apr_array_header_t *opt = apr_array_make(resultPool.getPool(),
                                           0, sizeof(const char*));

  if (flags & IGNORE_ALL_SPACE)
    APR_ARRAY_PUSH(opt, const char*) = "--ignore-all-space";
  if (flags & IGNORE_SPACE_CHANGE)
    APR_ARRAY_PUSH(opt, const char*) = "--ignore-space-change";
  if (flags & IGNORE_EOL_STYLE)
    APR_ARRAY_PUSH(opt, const char*) = "--ignore-eol-style";
  if (flags & SHOW_C_FUNCTION)
    APR_ARRAY_PUSH(opt, const char*) = "--show-c-function";

  /* TODO: Support -U (context size) */

  return opt;
}

svn_diff_file_options_t *DiffOptions::fileOptions(SVN::Pool &resultPool) const
{
  svn_diff_file_options_t *opt;

  opt = svn_diff_file_options_create(resultPool.getPool());

  if (flags & IGNORE_ALL_SPACE)
    opt->ignore_space = svn_diff_file_ignore_space_all;
  else if (flags & IGNORE_SPACE_CHANGE)
    opt->ignore_eol_style = svn_diff_file_ignore_space_change;

  if (flags & IGNORE_EOL_STYLE)
    opt->ignore_eol_style = TRUE;

  if (flags & SHOW_C_FUNCTION)
    opt->show_c_function = TRUE;

  /* TODO: Support context size */

  return opt;
}
